lib/common.js

```diff
@@ -156,50 +157,391 @@ HttpRequest.isBucketObject = function isBucketObject() {
 
 ///--- API
 
-function addCustomHeaders(req, res) {
+function addCustomHeaders(req, res, next) {
     var md = req.metadata.headers;
     var origin = req.headers.origin;
+    req.log.debug({
+        origin: origin,
+        hasMetadata: !!req.metadata,
+        metadataHeaders: md ? Object.keys(md) : 'undefined',
+        metadataHeadersRaw: md,
+        requestUrl: req.url,
+        requestMethod: req.method
+    }, 'CORS DEBUG: addCustomHeaders called');
+
+    // Separate CORS headers from other headers
+    var corsHeaders = {};
+    var regularHeaders = {};
+
+    if (md && Object.keys(md).length > 0) {
+        var mdKeys = Object.keys(md);
+        req.log.debug({
+            metadataKeysCount: mdKeys.length,
+            firstFewKeys: mdKeys.slice(0, 10)
+        }, 'CORS DEBUG: Processing metadata headers');
+
+        // Log each metadata key individually to see exact names
+        mdKeys.forEach(function (k) {
+            if (k.toLowerCase().includes('control') || k.toLowerCase().
+                includes('origin') || k.toLowerCase().includes('method')) {
+                req.log.debug({
+                    key: k,
+                    value: md[k],
+                    inCorsHeaders: CORS_RES_HDRS.indexOf(k) !== -1
+                }, 'CORS DEBUG: Potential CORS key');
+            }
+        });
+    }
 
-    Object.keys(md).forEach(function (k) {
-        var add = false;
-        var val = md[k];
-        // See http://www.w3.org/TR/cors/#resource-requests
-        if (origin && CORS_RES_HDRS.indexOf(k) !== -1) {
-            if (k === 'access-control-allow-origin') {
-                /* JSSTYLED */
-                if (val.split(/\s*,\s*/).some(function (v) {
-                    if (v === origin || v === '*') {
-                        val = origin;
-                        return (true);
+    Object.keys(md || {}).forEach(function (k) {
+        // Handle Manta's m- prefix for metadata headers
+        var headerName = k;
+        if (k.startsWith('m-')) {
+            headerName = k.substring(2); // Remove 'm-' prefix
+        }
+
+        var isCorsHeader = CORS_RES_HDRS.indexOf(headerName) !== -1;
```

Comment:
Is this why we always do cors processing elsewhere? We can catch/avoid it here if there are none?